h5py/_hl/attrs.py

```diff
@@ -95,13 +96,32 @@ def get_id(self, name):
 
     @with_phil
     def __setitem__(self, name, value):
-        """ Set a new attribute, overwriting any existing attribute.
+        """ Set a new attribute, overwriting or modifying any existing attribute.
+
+        Overwrite existing attributes if track_order is False, otherwise modify
+        attribute if type is preserved.
 
         The type and shape of the attribute are determined from the data.  To
         use a specific type or shape, or to preserve the type of an attribute,
         use the methods create() and modify().
         """
-        self.create(name, data=value)
+        with phil:
+            # see issue
+            # if attribute exists and track_order=True
+            if name in self and self._id.get_create_plist().get_attr_creation_order() > 0:
+                # extract type
+                if not isinstance(value, Empty):
+                    value = base.array_for_new_object(value, specified_dtype=None)
+                dtype = value.dtype
+                attr = h5a.open(self._id, self._e(name))
+                dt = attr.dtype
+                attr.close()
+                # if type is preserved -> modify
+                if dtype == dt:
+                    self.modify(name, value)
+                    return
```

Comment:
Or I guess we could implement it with delete+create, rather than create+delete+rename? It's less 'atomic-ish', but we already can't do an atomic replacement, and attributes are meant to be small, so maybe the simple thing is good enough.

Reply:
I hesitated to ask this, but delete+create would be the most user/downstream-friendly solution. I'm definitely in favour of that way.

Reply:
Looking back, I see the create+delete+rename mechanism was added in #501, which aimed to fix #498, but this looks orthogonal to the main fix there. I guess it was inspired by doing similar things with files, where we can atomically rename one file over another.

Given that we've now found a real case where rename is an issue, and attributes are meant to be small (so there's less risk of an error mid way through writing), I think it makes sense to go back to the simpler delete+create mechanism, even if that occasionally means you're left with neither after an error. Assuming it doesn't cause any test failures, that is. We could use this only when `track_order` is on, but I'm inclined to make it simpler rather than more complex.

@tacaswell @aragilar any objections?

Reply:
Yep, no objections from me.

Reply:
I'll try to free some time next week.

Reply:
Yes, feel free to take over. :heart: